keps/sig-api-machinery/2839-in-use-protection/README.md

```diff
@@ -0,0 +1,723 @@
+
+This KEP aims to provide a generic mechanism for protection that solves the above issues with Finalizers, because such use cases are also found for protecting secrets and configmaps in [KEP-2639](https://github.com/kubernetes/enhancements/pull/2640).
+
+Potential use cases are:
+- Protect `PersistentVolume` while being used by `PersistentVolumeClaim` (Replace the background logic for pv-protection)
+- Protect `PersistentVolumeClaim` while being used by `Pod` (Replace the background logic for pvc-protection)
+- Protect `Secret` while being used by `Pod`, `PersistentVolume`, `VolumeSnapshotContent` (Secret protection above)
+- Protect `ConfigMap` while being used by `Pod`
+- Protect `StorageClass` while being used by `PersistentVolume`
+- Protect `VolumeSnapshot` while being used by `VolumeSnapshotContent`
+- Protect `VolumeSnapshotClass` while being used by `VolumeSnapshotContent`
+- Protect `User`, `Group`, `ServiceAccount` and `Role` from `RoleBinding`
+- Protect `User`, `Group`, `ServiceAccount` and `ClusterRole` from `ClusterRoleBinding`
+- Protect resources that users marked as in-use
+
+### Goals
+
+- Provide a generic mechanism that prevents a resource from being deleted when it shouldn't be deleted
+
+### Non-Goals
+
+- Provide a specific mechanism to decide which particular objects should be prevented from being deleted when other particular objects exist
+- Provide a generic mechanism that prevents a resource from being updated when it shouldn't be updated
+
+## Proposal
+
+A new field `Liens` to mark the object not to be deleted is introduced in object Metadata.
+Deletion requests for resources with non-empty `Liens` will be blocked by a newly introduced validation in api-server.
+
+### User Stories (Optional)
+
+#### Story 1
+
+Another controller to protect Secret from deletion watches all Secrets and their potential user objects, like `Pod`, `PersistentVolumes`, and `VolumeSnapshotContent`. Once it find the Secret is used by one of the objects, it updates the Secret's `Liens` and ask this mechanism to block deletion request for the secret.
+
+#### Story 2
+
+A user sets `Liens` to an object to be protected by this mechanism. For example, when a user knows that `Namespace` X shouldn't be erroneously deleted while `Deployment` Y exists for his/her particular use case, the user can set such a protection.
+
+<!--
+### Notes/Constraints/Caveats (Optional)
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Most of the design ideas come from [here](https://github.com/kubernetes/kubernetes/issues/10179).
+Users or controllers can add a string to `Liens` to ask to protect a resource.
```

Comment:
Should we consider a subresource for this, so we can give tighter RBAC?

Reply:
I will check other KEPs on what should be needed to define this as a subresource and update the KEP (may be along with the final version of lien field).

Reply:
that could be something we evolve to for beta

Reply:
Added as a beta graduation criteria.

Reply:
Added as a beta graduation criteria.

Reply:
I think this should be a requirement for alpha so that we can get feedback about the actual API shape and so that clients will not have a guaranteed forced migration.  I've mentioned similar here: https://github.com/kubernetes/enhancements/pull/2840#issuecomment-1024238332